Automatic pairing in contact action

recuero opened the pull request:
Closes #24122.

recuero committed:
Set up contact pairs in contact action using sideset proximity (#24122)

A new input parameter (nodeset_list) is added to the contact action.
For (pseudo) axisymmetric solids whose contact interactions are consistently
defined by wrap-around sidesets, the contact action can now loop over the
nodes in a sideset and obtain an approximate "center of gravity".

Distances between centers of gravity are computed. A minimum and reference
distance is computed to identify sidesets whose interaction is likely. As such,
those sidesets are arbitrarily given the primary and secondary labels. After than,
we remove the possibly repeated pairs.

This feature may be useful in cases for which many axisymmetric bodies for contact
need to be set up. Picture a large number of hollow cylinders that can come
into contact.

recuero commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -901,6 +977,101 @@ ContactAction::addNodeFaceContact()
+            "geometry and the mesh exists. Deviations from axisymmetry may give rise to wrong "
+            "contact pairing. The user is encouraged to check the numerical results.");
+
+  _mesh = &_problem->mesh();
```
@lindsayad do we always get a serial mesh here? I am getting that the mesh is always serial despite using the `--distributed-mesh` option.

lindsayad replied:
Are you checking with `is_serial()`? At this stage of simulation setup, unless you've loaded a pre-split mesh, the mesh should indeed be serial. For any problems with displaced mesh, we delay remote element deletion until right before the `init_problem` task (which is close to the last task)

lindsayad replied:
If you're seeing that it's serial even during the simulation execution, then I would be pretty surprised. I would try rebasing on https://github.com/idaholab/moose/pull/24018 and see if that changes things

recuero replied:
Yes, checked with `is_serial()` and was only referring to this particular stage of system initialization. 

According to what you explained, if we wanted to make action decisions based on the mesh, it seems it'd be okay to do it at a stage such as the one in this PR and check that the mesh is serial to avoid issues with pre-split meshes (which are not being used much in contact problems at the moment)?

lindsayad replied:
Yes that sounds very reasonable to me

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1482242/) on b13d915 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24179/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1482256/) on b13d915 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>628ea9</th>
      <th colspan=4>#24179 b13d91</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.14%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>86877</td>
      <td>86879</td>
      <td>+2</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15160</td>
      <td>15159</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24179/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24179/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>628ea9</th>
      <th colspan=4>#24179 b13d91</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.04%</td>
      <td>90.11%</td>
      <td>+0.07%</td>
      <td>95.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3264</td>
      <td>3325</td>
      <td>+61</td>
      <td>76</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>361</td>
      <td>365</td>
      <td>+4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24179/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24179/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24179/coverage/xfem)

</details>



This comment will be updated on new commits.

recuero committed:
Fix logic bug in the generation of final pairs

lindsayad commented:
Unsubscribing for now. Ping me when ready for review

recuero committed:
Add test with a discarded contact pair

recuero commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1060,120 @@ ContactAction::addNodeFaceContact()
+      minimum_representative_distance = pair_distance.second;
+
+  // Apply a tolerance to the minimum distance to avoid missing pairs
+  minimum_representative_distance *= 1.2;
```
@bwspenc: Should we require user input here to gauge whether a reference distance would grant the generation of a new contact pair? I believe that for the case we have in mind is not strictly necessary, but may be good for other applications?

recuero replied:
Per our discussion, we will ask the user to provide a reference distance that will include some tolerance to capture computation error.

recuero committed:
Enhacements to automatic contact pair detection.

- Use a distance provided by the user as a threshold to generate contact pairs.
- Use areas, instead of node locations, to come up with the sideset's center of gravity

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1502319/) on bbfab96 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1502319/) on bbfab96 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1502319/) on bbfab96 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

bwspenc commented on modules/contact/include/actions/ContactAction.h:
```diff
@@ -99,6 +102,9 @@ class ContactAction : public Action
   /// Type that we use in Actions for declaring coupling
   typedef std::vector<VariableName> CoupledName;
 
+  /// Reference to mesh.
```
...except that it's a pointer 😄

bwspenc commented on modules/contact/include/actions/ContactAction.h:
```diff
@@ -113,7 +119,11 @@ class ContactAction : public Action
   void addContactPressureAuxKernel();
   /**
-   * Remove repeated contact pairs in the input file.
+   * Remove repeated contact pairs from boundary pairs.
```
I'd say 'from _boundary_pairs' to make it even more clear.

bwspenc commented on modules/contact/include/actions/ContactAction.h:
```diff
@@ -113,7 +119,11 @@ class ContactAction : public Action
    */
   void removeRepeatedPairs();
+  /**
+   * Create contact pairs from wrap-around sidesets for axisymmetric bodies.
```
More nit-picking about comments -- these don't have to be axisymmetric bodies. Maybe 'periodically repeating bodies'?

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -32,28 +37,44 @@ static unsigned int contact_userobject_counter = 0;
+  params.addParam<std::vector<BoundaryName>>(
       "secondary", "The list of boundary IDs referring to secondary sidesets");
+  params.addParam<std::vector<BoundaryName>>(
+      "sideset_list",
```
If we keep this capability in the standard contact action, I think the name of this parameter needs to convey what this does. Maybe 'automatic_pairing_boundaries'? I think that we pretty much universally use the term 'boundary' rather than 'sideset' in MOOSE, so it seems like we should use that (even though this boundary actually has to be a sideset in this case).

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -32,28 +37,44 @@ static unsigned int contact_userobject_counter = 0;
+      "neighbors. If this input is chosen to describe "
+      "contacting surfaces, separate 'primary' and 'secondary' inputs cannot be utilized. ");
+  params.addParam<Real>(
+      "automatic_pair_distance",
```
Suggested rename to: 'automatic_pairing_distance' to be consistent with the suggested name for the other parameter above.

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+ContactAction::createSidesetPairsFromGeometry()
+{
+  mooseInfo("The contact action is reading the list of sidesets and matching them assuming that: "
+            "1) Those sidesets wrap around the surfaces of the bodies, 2) Axisymmetry of the "
```
Is axisymmetry really the right term here? Seems like something like "periodically repeating geometry" is more appropriate.

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+  _mesh = &_problem->mesh();
+
+  if (!_mesh)
+    mooseError("Internal error in contact action.");
```
This is probably an error that will never be hit, but I'd still say what actually happened: "Failed to obtain mesh for automatically generating contact pairs"

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+    const auto find_set = sideset_ids.find(_mesh->getBoundaryID(sideset_name));
+    if (find_set == sideset_ids.end())
+      paramError("sideset_list",
+                 "The sideset ",
```
It might actually be a boundary that is in the mesh, but is not a sideset, so this message should be adjusted:
"<sideset_name> is not defined as a sideset in the mesh"
I'd drop the part about "Please revise your input file." because that goes without saying for any param error.

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -32,28 +37,44 @@ static unsigned int contact_userobject_counter = 0;
+      "bodies with straight extrusion axis (or a 2D tiled pattern) can be in contact with "
+      "neighbors. If this input is chosen to describe "
+      "contacting surfaces, separate 'primary' and 'secondary' inputs cannot be utilized. ");
+  params.addParam<Real>(
```
This should probably a range-checked parameter with a value >=0.

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+      }
+    }
+
+  mooseInfo("Reference distance between bodies for automatically assigning contact pairs in "
```
Printing this out doesn't really seem necessary -- we don't generally output other input parameters.

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1071,129 @@ ContactAction::addNodeFaceContact()
+  // and less than the threshold distance
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> lean_pairs_distances;
+  for (const auto & pair_distance : pairs_distances)
+    if (pair_distance.second > TOLERANCE &&
```
I'm not sure why we need the "> tolerance" check. Also, maybe use <= on the pair distance (although it probably doesn't really change much)? Finally, it seems a little weird to repeatedly get that parameter in a loop like this. We're probably not going to have enough sidesets for this to really be a performance issue, but it seems like it would still be better to get it once before the loop.

bwspenc reviewed (commented):
Also, regarding where this should live, I'm not really sure whether this belongs in the standard contact action, or maybe in a derived class, since this is a really specialized case. I was initially convinced that it should go in a derived class, but I'm starting to think it's fine how you have it as long as the parameters are appropriately named. I'm interested in what others think.

recuero committed:
Address Ben's comments

recuero committed:
Lowest boundary ID will be secondary

lindsayad commented on framework/src/base/Moose.C:
```diff
@@ -190,6 +190,7 @@ addActionTypes(Syntax & syntax)
   registerTask("prepare_mesh", false);
   registerTask("delete_remote_elements_after_late_geometric_ghosting", false);
   registerTask("setup_mesh_complete", true); // calls prepare
+  registerTask("check_generated_mesh", false);
```
I am ok with this but I would rename it to `post_mesh_prepared`

lindsayad commented on framework/src/base/Moose.C:
```diff
@@ -280,6 +281,7 @@ addActionTypes(Syntax & syntax)
                            "(create_problem_custom)"
                            "(create_problem_default)"
                            "(create_problem_complete)"
+                           "(check_generated_mesh)"
```
With its new `post_mesh_prepared` name, let's move this to right after `setup_mesh_complete`

recuero replied:
Please see my last commit. CIVET was crashing without that last revert commit.

lindsayad replied:
I think we need to understand why it was crashing. If the task relies on something more than the mesh being prepared, then we need to rename it

lindsayad replied:
Oh I see the issue. You're calling `_problem->mesh()`. In an action you should already have access to `_mesh` and `_displaced_mesh` members. You don't need to go through the `_problem`

recuero replied:
Ah, that makes sense. Will make that change.

recuero committed:
Address Alex's review: 'post_mesh_prepared' initialization stage

recuero committed:
No need to get mesh through problem (Alex's review)

bwspenc reviewed (changes requested):
Just a few minor things to address. Otherwise, looks good.

Also, this seems like a major enough feature that we should document this with a few sentences at least in the .md file for the action.

recuero commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1073,166 @@ ContactAction::addNodeFaceContact()
+ContactAction::createSidesetPairsFromGeometry()
+{
+  mooseInfo("The contact action is reading the list of boundaries and automatically matching them "
+            "assuming that: 1) Those boundaries wrap around the surfaces of the bodies, 2) "
```
Yes, let's shift that responsibility to the user.

recuero commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1073,166 @@ ContactAction::addNodeFaceContact()
+  {
+    if (pair_distance.second > TOLERANCE && pair_distance.second <= automatic_pairing_distance)
+      lean_pairs_distances.emplace_back(pair_distance);
+    else if (pair_distance.second <= TOLERANCE)
```
I did this per your request!

bwspenc replied:
I requested this change mostly to improve on the loop structure in the original version of this code. Then upon giving this some thought, I realized that there really are good use cases for the case where the distance is 0. I regret the inconvenience I caused by changing requests, but iterative changes in direction are inherent in software development. 

I would really prefer that we get rid of this use of tolerance (and the warning associated with it) entirely because it can give us false positives about potential issues. Legitimate uses of this are still going to trigger these warnings.

To check for this particular user error, it would be better to go through each primary/secondary pair face by face to make sure they're not identical. I guess while we're at it, we really should be more stringent and make sure that a given face does not appear in both the primary and secondary boundaries. That check should be done for all uses of the contact action, not just for this automatic pairing, because that is not an issue specific to this feature. The downsides to doing that is that it doesn't work with pre-split meshes, and it could be slow (but probably not all that slow).

I think it's a good idea to do that sometime, but in a different PR. Let's leave this so it generates a warning now, and then remove that warning later when we implement the more rigorous check.

recuero committed:
Address Ben's comments

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -1028,6 +1073,143 @@ ContactAction::addNodeFaceContact()
+                        ", with a relative distance of ",
+                        pair_distance.second);
+      mooseWarning("Distance between contact pairs is small");
+    }
```
Per our discussion, let's just drop the tolerance and associated warning entirely, which makes this even simpler.

recuero committed:
Remove minimum distance (tolerance) to match boundaries

bwspenc reviewed (approved):
Looks great, thanks for all the iterative updates to this!

lindsayad reviewed (commented):
Is this something we should put in our newsletter?

recuero commented:
This is a fairly specialized capability, but we could add a sentence about it (is there a PR on it?)

We should definitely add the new penalty/AL mortar capabilities.

lindsayad commented:
We've taken to a workflow where items get added to the newsletter in the PR that they're introduced in. If you don't think this is a big deal, don't worry about it

recuero commented:
Okay, I'll remember to suggest to include it in the PR with the AL implementation or right after it.

recuero commented:
@lindsayad are you okay with this PR?

lindsayad reviewed (approved):
yep!

recuero commented:
Thanks! I'll test this one more time in a large-scale problem and will merge after that.

recuero commented on modules/contact/test/tests/multiple_contact_pairs/three_hexagons_coarse_automatic_pair.i:
```diff
@@ -0,0 +1,117 @@
+    model = frictionless
+    penalty = 2e+03
+    normalize_penalty = true
+    automatic_pairing_distance = 2.75
```
`automatic_pairing_boundaries = '1 2 3 4 ... 144'`
